### PR TITLE
2D U-Net HipMRI s46978116

### PR DESCRIPTION
**Pull request by obbamna**
### 2D U-Net Solution to Segmenting HipMRI Data

I am requesting to merge my solution to the 2D U-Net problem in COMP3710 Pattern Recognition Report. I implemented and trained a 2D U-Net to segmeent HipMRI data based on body parts. I have saved two models as well as various images regarding their performances. README.MD contains a report of the 2D U-Net's implementation and results.

### Files

My pull request contains these files:

- **dataset.py**

Provides functions and classes to load images from the HipMRI dataset found on RANGPUR. Replace folder path with your own path where necessary.

- **modules.py**

Contains the Neural Networks that makes up the 2D U-Net

- **train.py**

Trains a 2D U-Net from modules.py on data using the dataset class from dataset.py to load images in tensor form. Runs training for a number of epochs and generates and saves prediction images, training plots and loss performances. 

- **predict.py**

Runs a pre-trained model on the test set to evaluate it. Produces prediction images and dice co-efficients on test set.

- **README.MD**

Contains a report of the entire project with output, theory, requirements and references.

- **images/**

Folder containing images used in the report and the saved dice co-efficients.

- **savedmodels/**

Folder containing two saved models that can be evaluated in predict.py

### Not Included

HipMRI data has not been included, however it is necessary for training and evaluating and must be acquired independently. Replace necessary folder paths with your own path to this data.

### CODE REVIEW

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**  
1. **Solves problem**: 4 labels achieved over 0.75 DICE score, 2 labels achieved nearly 0.75 DICE score. The tricks and experiments used for improving accuracy were well-documented in the readme. The loss plots and prediction figures can be found under the "images" folder; the image paths in the readme file need to be corrected.  Overall, the problem is well-solved.
2. **Implementation**: Some [hypterparameters](https://github.com/obbamna/PatternAnalysis-2024/blob/topic-recognition/recognition/prostate2dnet_46978116/train.py#L18C1-L26C4) were hard-coded in the training script instead of using ArgParse. Otherwise all great.
3. **Design**: well-designed.
4. **Comment**: sufficient comments.
5. **Difficulty**: **Easy Difficulty**.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Easy. UNet2D (Easy Difficulty)</td><td>-10</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. Spans 2 days.</td><td>-2</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. No train graphs.</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Pull Request has problems: Late submission.</td><td>-2</td></tr>
	<tr><td></td><td>Feedback action required: Feedback marks possible +2 if the requested changes are made. Remove model checkpoints for mearge.</td><td>-2</td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-18</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by obbamna** [comment]
Requested changes from feedback have been filled.

I would really appreciate leniency on the -2 for late submission as I had a one week extension that was approved and the pull request was submitted before that extended due date (even earlier if you count the Monday pseudo-deadline). However, I understand if I still lose marks on late feedback.

Also, there are training graphs in images/graphs. 

Thank you.

**Comment by hanemma7moud** [comment]
Mitchell [Flaherty](https://learn.uq.edu.au/webapps/gradebook/do/instructor/enterGradeCenter?course_id=_182685_1&cvid=fullGC#)

**Comment by obbamna** [comment]
Pull request wasn't late and feedback was applied and there are training graphs. Missing 5 Marks.

**Comment by shakes76** [comment]
Approved extension +2, feedback applied + 2
